receiver/sapnetweaverreceiver/scraper.go

```diff
@@ -223,3 +229,75 @@ func (s *sapNetweaverScraper) collectGetAlertTree(_ context.Context, now pcommon
 	s.recordSapnetweaverSessionsBrowserCountDataPoint(now, alertTreeResponse, errs)
 	s.recordSapnetweaverSessionsEjbCountDataPoint(now, alertTreeResponse, errs)
 }
+
+// collectCertificateValidity collects certificate final validity date.
+func (s *sapNetweaverScraper) collectCertificateValidity(_ context.Context, now pcommon.Timestamp, errs *scrapererror.ScrapeErrors) {
+	args := []string{"-L", "/usr/sap", "-name", "*.pse"}
+	certFilesPaths, err := s.service.FindFile(args...)
+	if err != nil {
+		errs.AddPartial(1, fmt.Errorf("failed to find certificate files: %w", err))
+		return
+	}
+
+	for _, certFilePath := range certFilesPaths {
+		if certFilePath == "" {
+			continue
+		}
+
+		command := fmt.Sprintf("/usr/sap/hostctrl/exe/sapgenpse get_my_name -p %s -n validity", certFilePath)
+		certs, err := s.service.CertExecute(command)
+		if err != nil {
+			errs.AddPartial(1, fmt.Errorf("failed to execute certificate at: %s, error: %w", certFilePath, err))
+			continue
+		}
+
+		for _, line := range certs {
+			if strings.Contains(line, "NotAfter") {
+				// extracts timestamp within the parenthesis like: NotAfter : Thu Dec 31 19:00:01 2037 (380101000001Z)
+				re := regexp.MustCompile(`\((.*?)\)`)
```

Comment:
Can we move this regex out of the function to the top level?

Reply:
moved regex out of loops